src/main/java/com/linagora/james/mailets/ExtractPhoneNumberMailet.java

```diff
@@ -0,0 +1,151 @@
+
+    static final String HEADER_NAME = "headerName";
+    static final String HEADER_NAME_DEFAULT_VALUE = "X-Phone-Number";
+
+    @VisibleForTesting String headerName;
+    private final UUIDGenerator uuidGenerator;
+    private final PhoneNumberUtil phoneNumberUtil;
+
+    private static final MimeConfig MIME_ENTITY_CONFIG = MimeConfig.custom()
+        .setMaxContentLen(-1)
+        .setMaxHeaderCount(-1)
+        .setMaxHeaderLen(-1)
+        .setMaxHeaderCount(-1)
+        .setMaxLineLen(-1)
+        .build();
+
+    public ExtractPhoneNumberMailet() {
+        this(new UUIDGenerator());
+    }
+
+    @VisibleForTesting
+    ExtractPhoneNumberMailet(UUIDGenerator uuidGenerator) {
+        this.uuidGenerator = uuidGenerator;
+        this.phoneNumberUtil = PhoneNumberUtil.getInstance();
+    }
+
+    @Override
+    public void init() throws MessagingException {
+        headerName = getInitParameter(HEADER_NAME, HEADER_NAME_DEFAULT_VALUE);
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("headerName value: " + headerName);
+        }
+        if (Strings.isNullOrEmpty(headerName)) {
+            throw new MailetException("'headerName' is mandatory");
+        }
+    }
+
+    @Override
+    public String getMailetInfo() {
+        return "ExtractPhoneNumberMailet";
+    }
+
+    @Override
+    public void service(Mail mail) {
+        try {
+            MessageContentExtractor messageContentExtractor = new MessageContentExtractor();
+            MessageContentExtractor.MessageContent messageContent = messageContentExtractor.extract(parse(mail));
+            Optional<String> textBody = messageContent.getTextBody();
+
+            textBody.flatMap(this::extractPhoneNumber).ifPresent(phoneNumber -> {
+                try {
+                    mail.getMessage().addHeader(headerName, phoneNumber);
+                } catch (MessagingException e) {
+                    LOGGER.error("Exception on ExtractPhoneNumberMailet", e);
+                }
+            });
+
+        } catch(Exception e) {
+            LOGGER.error("Exception on ExtractPhoneNumberMailet", e);
+        }
+    }
+
+    private org.apache.james.mime4j.dom.Message parse(Mail mail) throws MailetException {
+        try {
+            return org.apache.james.mime4j.dom.Message.Builder
+                .of()
+                .use(MIME_ENTITY_CONFIG)
+                .parse(mail.getMessage().getInputStream())
+                .build();
+        } catch (MessagingException | IOException e) {
+            throw new MailetException("Unable to parse message: " + e.getMessage(), e);
+        }
+    }
+
+    @VisibleForTesting
+    Optional<String> extractPhoneNumber(String text) {
+        List<PhoneNumberMatch> numbers = ImmutableList.copyOf(phoneNumberUtil.findNumbers(text, "fr"));
+
+        if (numbers.size() == 0) {
+            return Optional.empty();
+        }
+
+        return Optional.of(numbers.get(numbers.size() - 1).rawString());
```

Comment:
Iterables.getLast ?

And why not keeping all phone numbers, and add them to headers?